src/main/java/net/fabricmc/loader/impl/launch/FabricMixinBootstrap.java

```diff
@@ -79,7 +75,79 @@ public static void init(EnvType side, FabricLoaderImpl loader) {
+
+		for (ModContainerImpl mod : loader.getModsInternal()) {
+			for (String config : mod.getMetadata().getMixinConfigs(side)) {
+				ModContainerImpl prev = configToModMap.putIfAbsent(config, mod);
+				if (prev != null) throw new RuntimeException(String.format("Non-unique mixin config name %s used by %s and %s", config, prev.getMetadata().getId(), mod.getMetadata().getId()));
+
+				Mixins.addConfiguration(config);
+			}
+		}
+
+		for (Config config : Mixins.getConfigs()) {
+			ModContainerImpl mod = configToModMap.get(config.getName());
+			if (mod == null) continue;
+		}
+
+		try {
+			IMixinConfig.class.getMethod("decorate", String.class, Object.class);
+			MixinConfigDecorator.apply(configToModMap);
+		} catch (NoSuchMethodException e) {
+			Log.info(LogCategory.MIXIN, "Detected old Mixin version without config decoration support");
+		}
+
 		initialized = true;
 	}
+
+	private static final class MixinConfigDecorator {
+		private static final SemanticVersion LAST_LOADER_MIXIN_0_9_2 = new SemanticVersionImpl(new int[] { 0, 11, 9999 }, null, null);
+
+		static void apply(Map<String, ModContainerImpl> configToModMap) {
+			for (Config rawConfig : Mixins.getConfigs()) {
+				ModContainerImpl mod = configToModMap.get(rawConfig.getName());
+				if (mod == null) continue;
+
+				IMixinConfig config = rawConfig.getConfig();
+				config.decorate(FabricUtil.KEY_MOD_ID, mod.getMetadata().getId());
+				config.decorate(FabricUtil.KEY_COMPATIBILITY, getMixinCompat(mod));
+			}
+		}
+
+		private static int getMixinCompat(ModContainerImpl mod) {
+			Manifest manifest = FabricLauncherBase.getLauncher().getManifest(mod.getOriginPath());
+			String versionStr;
+
+			if (manifest != null
+					&& "net.fabricmc".equals(ManifestUtil.getManifestValue(manifest, ManifestUtil.NAME_MIXIN_GROUP))
+					&& (versionStr = ManifestUtil.getManifestValue(manifest, ManifestUtil.NAME_MIXIN_VERSION)) != null) {
+				try {
+					SemanticVersion version = SemanticVersion.parse(versionStr);
+					int major = version.getVersionComponent(0);
+					int minor = Math.min(version.getVersionComponent(1), 999);
+					int patch = Math.min(version.getVersionComponent(2), 999);
+
+					return (major * 1000 + minor) * 1000 + patch;
+				} catch (VersionParsingException e) {
+					Log.warn(LogCategory.GENERAL, "Error parsing Mixin Version from Manifest for %s", mod, e);
+				}
+			}
+
+			// no manifest record, try to infer from loader dependency being >= 0.12.x (first with fabrix-mixin 0.9.4+)
+
+			for (ModDependency dep : mod.getMetadata().getDependencies()) {
+				if (dep.getKind() == Kind.DEPENDS && (dep.getModId().equals("fabricloader") || dep.getModId().equals("fabric-loader"))) {
+					if (!dep.matches(LAST_LOADER_MIXIN_0_9_2)) { // not satisfied by a loader version with the old mixin version
+						return FabricUtil.COMPATIBILITY_0_9_4;
+					}
+				}
+			}
+
+			// default to old fabric-mixin
+
+			return FabricUtil.COMPATIBILITY_0_9_2;
```

Comment:
In dev this should latest when a classpath mod, but fallback to to older version when loading mods from the mods dir.

Reply:
This might cause issues with outdated deps though... not sure there is a way to handle all the cases here.